crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -37,68 +51,101 @@ impl CoinGecko {
+                "failed to sent CoinGecko price request: {e:?}"
+            ))
+        })?;
+        if !response.status().is_success() {
+            let status = response.status();
+            return match status {
+                StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
+                status => Err(PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to retrieve prices from CoinGecko: error with status code {status}."
+                ))),
+            };
+        }
+        let response = response.text().await;
+        observe::coingecko_response(&url, response.as_deref());
+        let response = response.map_err(|e| {
+            PriceEstimationError::EstimatorInternal(anyhow!(
+                "failed to fetch native CoinGecko prices: {e:?}"
+            ))
+        })?;
+        let prices = serde_json::from_str::<Response>(&response)
+            .map_err(|e| {
+                PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to parse native CoinGecko prices from {response:?}: {e:?}"
+                ))
+            })?
+            .0;
+
+        let price = prices
+            .get(&token)
+            .ok_or(PriceEstimationError::NoLiquidity)?;
+        Ok(price.eth)
+    }
 }
 
 impl NativePriceEstimating for CoinGecko {
     fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
         async move {
-            let mut url = crate::url::join(&self.base_url, &self.chain);
-            url.query_pairs_mut()
-                .append_pair("contract_addresses", &format!("{:#x}", token))
-                .append_pair("vs_currencies", "eth")
-                .append_pair("precision", "full");
-            let mut builder = self.client.get(url.clone());
-            if let Some(ref api_key) = self.api_key {
-                builder = builder.header(Self::AUTHORIZATION, api_key)
-            }
-            observe::coingecko_request(&url);
-            let response = builder.send().await.map_err(|e| {
-                PriceEstimationError::EstimatorInternal(anyhow!(
-                    "failed to sent CoinGecko price request: {e:?}"
-                ))
-            })?;
-            if !response.status().is_success() {
-                let status = response.status();
-                return match status {
-                    StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
-                    status => Err(PriceEstimationError::EstimatorInternal(anyhow!(
-                        "failed to retrieve prices from CoinGecko: error with status code \
-                         {status}."
-                    ))),
-                };
+            match self.native_price {
+                NativePrice::Eth => self.send_request_price_in_eth(token).await,
+                NativePrice::Other(native_price_token) => {
+                    let token_eth = rust_decimal::Decimal::try_from(
```

Comment:
We aren't caching here since egress-proxy is caching. My idea is to see the performance of it and how many times the cache is `hit` or `miss`, and then assess if it makes sense or not implement internal cache for the `wxDai` prices.